io_u.c

```diff
@@ -2078,12 +2079,37 @@ static void file_log_write_comp(const struct thread_data *td, struct fio_file *f
 	if (!f->last_write_comp)
 		return;
 
+	now_nsec = utime_since_now(&io_u->start_time) * 1000;  /* Convert to nanoseconds */
+
 	idx = f->last_write_idx++;
-	f->last_write_comp[idx] = offset;
+	f->last_write_comp[idx].offset = offset;
+	f->last_write_comp[idx].completion_time_nsec = now_nsec;
+	f->last_write_comp[idx].flags = 0;
+	f->last_write_comp[idx].flush_count = f->flush_count;
+
+	/* Check if this is a FUA write */
+	if (io_u && (io_u->flags & IO_U_F_FUA))
+		f->last_write_comp[idx].flags |= FIO_WRITE_COMP_FUA;
+
 	if (f->last_write_idx == td->last_write_comp_depth)
 		f->last_write_idx = 0;
 }
 
+static void file_log_flush_comp(struct fio_file *f, struct io_u *io_u)
+{
+	uint64_t completion_time;
+
+	if (!f)
+		return;
+
+	/* Track the last FLUSH completion timestamp */
+	completion_time = utime_since_now(&io_u->start_time) * 1000;
```

Comment:
`ntime_since_now`
Also, we already calculate a completion time elsewhere for all operations. Is there a way to use that completion time instead of calculating a second one?